client/src/worker_registry/mod.rs

```diff
@@ -0,0 +1,288 @@
+//! This is needed to implement Eager Workflow Start, a latency optimization in which the client,
+//!  after reserving a slot, directly forwards a WFT to a local worker.
+
+use rand::{seq::SliceRandom, thread_rng};
+use std::{collections::HashMap, fmt::Debug};
+use temporal_sdk_core_protos::temporal::api::workflowservice::v1::PollWorkflowTaskQueueResponse;
+
+/// This trait wraps a worker, which provides WFT processing slots.
+#[cfg_attr(test, mockall::automock)]
+pub trait SlotProvider: Send + Sync + Debug {
+    /// A unique identifier for the worker.
+    fn uuid(&self) -> String;
+    /// The namespace for the WFTs that it can process.
+    fn namespace(&self) -> String;
+    /// The task queue this provider listens to.
+    fn task_queue(&self) -> String;
+    /// Try to reserve a slot with matching namespace and task queue.
+    fn try_reserve_wft_slot(&self) -> Option<Box<dyn Slot>>;
+}
+
+/// This trait represents a slot reserved for processing a WFT by a worker.
+#[cfg_attr(test, mockall::automock)]
+pub trait Slot: Send + Sync {
+    /// Consumes this slot by dispatching a WFT to its worker. This can only be called once.
+    fn schedule_wft(&mut self, task: PollWorkflowTaskQueueResponse) -> Result<(), anyhow::Error>;
+}
+
+/// This trait enables local workers to made themselves visible to a shared client instance.
+pub trait WorkerRegistry: Send + Sync {
+    /// Register a local worker that can provide WFT processing slots.
+    fn register(&mut self, provider: Box<dyn SlotProvider>);
+    /// Unregister a provider, typically when its worker starts shutdown.
+    fn unregister(&mut self, uuid: String);
+}
+
+/// Implements a [WorkerRegistry] and provides a convenient method
+/// to find compatible slots within the collection.
+#[derive(Default, Debug)]
+pub struct SlotManager {
+    /// Maps keys, i.e., namespace#task_queue, to providers.
+    providers: HashMap<String, Vec<Box<dyn SlotProvider>>>,
+    /// Maps uuids to keys in `providers`.
+    index: HashMap<String, String>,
+}
+
+fn to_key(namespace: &str, task_queue: &str) -> String {
+    format!("{}#{}", namespace, task_queue)
+}
+
+fn random_permutation(num_entries: usize) -> Vec<usize> {
+    let mut vec: Vec<usize> = (0..num_entries).collect();
+    vec.shuffle(&mut thread_rng());
+    vec
+}
+
+impl SlotManager {
+    /// Factory method.
+    pub fn new() -> Self {
+        Self {
+            index: HashMap::new(),
+            providers: HashMap::new(),
+        }
+    }
+
+    /// Try to reserve a compatible processing slot in any of the registered workers.
+    pub fn try_reserve_wft_slot(
+        &self,
+        namespace: String,
+        task_queue: String,
+    ) -> Option<Box<dyn Slot>> {
+        let key = to_key(&namespace, &task_queue);
+        if let Some(all) = self.providers.get(&key) {
+            let index = random_permutation(all.len());
+            for i in &index {
```

Comment:
You don't need this permuting helper, you can use: https://docs.rs/rand/latest/rand/seq/trait.IteratorRandom.html#method.choose

Reply:
But, that said, I kinda question if we need any of this anyway. IMO I am 100% fine with the restriction that you can't have more than one worker pulling from the same task queue in the same process. That just doesn't make much sense to me.

Reply:
That also allows us to get rid of the UUID per-worker, which would be nice to not have I think.

Reply:
This is coming from the Go and Java implementations where you can have multiple workers to the same queue+namespace and you want to uniformly distribute the load with randomness.
I'm not sure we can always guarantee that there is only one worker per task queue for all core languages, and uniform distribution helps not only with load but also robustness...
 And if one day we will fix versioning, it may be easier if we already support many to one...

Reply:
I just checked the IteratorRandom but it doesn't give you a permutation, i.e., it can return the same element many times and miss another one for a very long time... The `choose_multiple` method  seems to have this problem too, it also says that for fully random  order you need to shuffle the result, which kind of defeats the purpose...

Reply:
Yeah, I just don't think we need multiple workers on the same queue in the same process. In fact, that's illegal already in Java: https://github.com/temporalio/sdk-java/blob/8af4a2647dacab2cb0d4a3a613482e00bc297a39/temporal-sdk/src/main/java/io/temporal/worker/WorkerFactory.java#L167

So, let's avoid some complexity and forbid it. I can't think of any reasonable use-case.

Reply:
Ok, that simplifies a lot of things. I'll just warn with ignored conflict registrations, I don't want to break existing code in the unlikely event they use multiple workers/queue in the same process...